### PR TITLE
Issue219 flag control compute custom only

### PR DESCRIPTION
**Pull request by jservonnat**

### CODE REVIEW

**Comment by jservonnat** [comment]
@doutriaux1 PR on the change to compute only the custom metrics:
- by default, compute only pcmdi metrics
- if compute_custom, compute both
- if compute_custom and compute_custom_only='true', compute only the custom metrics
Getting close to my first PR to the PCMDI MP :)?

**Comment by durack1** [comment]
@jservonnat great you're up.. The param file in particular looks good, you've certainly made the most of functionality available within the package!

**Comment by doutriaux1** [comment]
@jservonnat there's some issues here. Why do you need to edit anything in our repo? Can't you control everything from the parameter file?

File: src/python/pcmdi/__init__.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
 import io
 from mean_climate_metrics_calculations import compute_metrics
+from mean_climate_metrics_calculations_ExtInfo import compute_metrics_ExtInfo
```
Comment (doutriaux1) [inline src/python/pcmdi/__init__.py]: can't let that in our repo will break evry system that doesn't have your stuff this probably needs to be done in the parameter file

Comment (doutriaux1) [inline src/python/pcmdi/__init__.py]: @durack1 @gleckler1 do we want these calculation as part of our distribution? Or should this only be installed at IPSL where they need it?

File: src/python/pcmdi/mean_climate_metrics_calculations_ExtInfo.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+import cdms2 as cdms
```
Comment (doutriaux1) [inline src/python/pcmdi/mean_climate_metrics_calculations_ExtInfo.py]: @gleckler1 @durack1 same question again. Do we want these calaculations as part of our repo?

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/home/igcmg/PCMDI-MP/devel_install_PCMDI_IPSL/PCMDI_METRICS/bin/python
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: NOOOOOOOO! This line will break every systems!

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 4]
```diff
@@ -288,9 +288,13 @@ def applyCustomKeys(O,custom_dict,var):
                          Vr=fv[varInFile]
                          ## Need to recover only first time/leve/etc...
                          N=Vr.rank()-2 # minus lat/lon
+                         #<<<<<<< HEAD
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: these look like left over from a bad merge... worrisome...

File: src/python/pcmdi/__init__.py (hunk 1 above)
Comment (jservonnat) [inline src/python/pcmdi/__init__.py]: @doutriaux1 This is a mistake... sorry it's my first PR... The "Ext_Info" stuff has no reason to be here anymore. I don't even use it. If you prefer I can set up a new PR.

File: src/python/pcmdi/mean_climate_metrics_calculations_ExtInfo.py (hunk 2 above)
Comment (jservonnat) [inline src/python/pcmdi/mean_climate_metrics_calculations_ExtInfo.py]: @doutriaux1 Same...

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py (hunk 3 above)
Comment (jservonnat) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: @doutriaux1 I don't even know why this line is here...
After looking at the driver in the branch, I really think I've done something wrong. I don't even see the modifications to compute only the custom metrics... I'm not a git master yet...
I'll check what I've done and get back to you with another PR.

**Comment by jservonnat** [comment]
@doutriaux1 As an answer to your previous question, the problem is that you can control the computation of the custom metrics from the parameter file (as it is in the pcmdi master) but you will automatically compute the pcmdi metrics at the same time. This modification aims at computing only the custom metrics.

**Comment by jservonnat** [comment]
I've set up a new PR with the modifications... @doutriaux1 can you have a look at this one and say if it's ok?

**Comment by durack1** [comment]
@doutriaux1 this PR is up to you..

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 5]
```diff
@@ -108,12 +102,14 @@ def applyCustomKeys(O,custom_dict,var):
     sftlf[model_version]["md5"]=None
 if parameters.targetGrid == "2.5x2.5":
   tGrid = cdms2.createUniformGrid(-88.875,72,2.5,0,144,2.5)
-else:
+
+if parameters.targetGrid == "1.0x1.0":
+  tGrid = cdms2.createUniformGrid(-89,179,1,0,360,1)
+
+if parameters.targetGrid not in ['2.5x2.5','1.0x1.0']:
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: can we use if/elif/else here?

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 6]
```diff
@@ -108,12 +102,14 @@ def applyCustomKeys(O,custom_dict,var):
...
   tGrid = parameters.targetGrid
 
-sft = cdutil.generateLandSeaMask(tGrid)
-sft[:]=sft.filled(1.)*100.
-sftlf["targetGrid"] = sft
+sftlf["targetGrid"] = cdutil.generateLandSeaMask(tGrid)*100.
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: why removing the filled(1.) I think it is needed in some case land comes back as missing

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 7]
```diff
@@ -159,13 +155,12 @@ def applyCustomKeys(O,custom_dict,var):
...
-    if isinstance(refs,list) and "all" in [x.lower() for x in refs]:
-      refs = "all"
-    if isinstance(refs,(unicode,str)):
+    dup('ref is: ',parameters.ref)
+    if isinstance(parameters.ref,list):
+        refs=parameters.ref
+    elif isinstance(parameters.ref,(unicode,str)):
         #Is it "all"
-        if refs.lower()=="all":
+        if parameters.ref.lower()=="all":
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: this bits worries me a lot, why not using refs any longer, are you positive it's not used anywhere else?

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 8]
```diff
@@ -266,10 +259,14 @@ def applyCustomKeys(O,custom_dict,var):
                   MODEL.realm = realm
                   MODEL.period = parameters.period  
                   MODEL.ext="nc"
+		  print 'targetGrid ='
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: please remove or use dup()

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 9]
```diff
@@ -283,14 +280,16 @@ def applyCustomKeys(O,custom_dict,var):
                        #ok we can try to generate the sftlf
                        MODEL.variable=var
                        dup("auto generating sftlf for model %s " % MODEL())
+		       print 'last check 1'
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: yes please remove debug statements

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 10]
```diff
@@ -283,14 +280,16 @@ def applyCustomKeys(O,custom_dict,var):
...
                          Vr=fv[varInFile]
+			 print 'shape Vr'
+			 print Vr.shape
                          ## Need to recover only first time/leve/etc...
                          N=Vr.rank()-2 # minus lat/lon
-                         sft = cdutil.generateLandSeaMask(Vr(*(slice(0,1),)*N))*100.
-                         sft[:]=sft.filled(100.)
-                         sftlf[model_version]["raw"]=sft
+                         sftlf[model_version]["raw"]=cdutil.generateLandSeaMask(Vr(*(slice(0,1),)*N))*100.
+		         print 'last check 2'
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: same comment here the filled is here for a reason, please put back

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 11]
```diff
@@ -300,9 +299,9 @@ def applyCustomKeys(O,custom_dict,var):
                      if level is None:
                        OUT.level=""
                        dm = MODEL.get(var,varInFile=varInFile)  #+"_ac")
+		       print 'last check'
                      else:
                        OUT.level = "-%i" % (int(level/100.))
-                       #Ok now fetch this
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: why removing?

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 12]
```diff
@@ -300,9 +299,9 @@ def applyCustomKeys(O,custom_dict,var):
                      if level is None:
                        OUT.level=""
                        dm = MODEL.get(var,varInFile=varInFile)  #+"_ac")
+		       print 'last check'
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: debug statement

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 13]
```diff
@@ -345,7 +344,7 @@ def applyCustomKeys(O,custom_dict,var):
                               "ModellingGroup":"institute_id",
                               "Experiment":"experiment",
                               "ModelFreeSpace":"ModelFreeSpace",
-                              "Realization":"realization",
+                              "SimName":"realization",
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: did we agree on this?

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 14]
```diff
@@ -373,7 +372,8 @@ def applyCustomKeys(O,custom_dict,var):
                                         vals.append("N/A")
                                     f.close()
                             descr[att] = fmt % tuple(vals)
-                      metrics_dictionary[model_version]["SimulationDescription"] = descr 
+                      #metrics_dictionary[model_version]["SimulationDescription"] = descr
+                      metrics_dictionary[model_version]["SimulationDescription"] = descr
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: ??? commented and back on?

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 15]
```diff
@@ -383,14 +383,35 @@ def applyCustomKeys(O,custom_dict,var):
...
                   if not metrics_dictionary[model_version].has_key(refabbv):
                     metrics_dictionary[model_version][refabbv] = {'source':onm}
+
                   pr = metrics_dictionary[model_version][refabbv].get(parameters.realization,{})
-                  pr_rgn = pcmdi_metrics.pcmdi.compute_metrics(var,dm,do)
+                  
+                  pr_rgn = {}
+                  if hasattr(parameters,"compute_custom_metrics")==False:
+		     pr_rgn.update(pcmdi_metrics.pcmdi.compute_metrics(var,dm,do))
+		  else:
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: else seems out of alignement

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 16]
```diff
@@ -383,14 +383,35 @@ def applyCustomKeys(O,custom_dict,var):
...
+                     if hasattr(parameters,"compute_custom_only"):
+                        compute_custom_only = parameters.compute_custom_only
+		     else:
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: same here! Does this actually run?

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 17]
```diff
@@ -383,14 +383,35 @@ def applyCustomKeys(O,custom_dict,var):
...
+		        compute_custom_only='false'
+			print "set compute_custom_only='true' in the param file to compute only the custom metrics"
+		     print 'compute_custom_only = ',compute_custom_only
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: debug statement

File: src/python/pcmdi/scripts/pcmdi_metrics_driver.py [hunk 18]
```diff
@@ -383,14 +383,35 @@ def applyCustomKeys(O,custom_dict,var):
...
+                     
+		     if compute_custom_only=='false':
```
Comment (doutriaux1) [inline src/python/pcmdi/scripts/pcmdi_metrics_driver.py]: use is not == for False/True it's safer

actually this should be:
```
if not compute_custom_only:
```

**Comment by doutriaux1** [comment]
please address comments and add a test case

**Comment by durack1** [comment]
@jservonnat we'll close this PR in the coming days.. Did you have an issue with that?

**Comment by jservonnat** [comment]
@durack1 @doutriaux1 sorry guys, I've been away for a while... I'll do my best to sort those things out this week.
I'm using this flag a lot for what we do at IPSL. This is very convenient to compute only the custom metrics, without recalculating the PCMDI metrics. Also it allows having one parameter file by set of custom metrics. This way I can better organize my workflow.
I'll get back to you as soon as possible.

**Comment by jservonnat** [comment]
@doutriaux1 @durack1 @gleckler1 Ok I've finally managed to install the package again. The problem was apparently coming from the install_metrics.bash script.
The question is: are you guys interested in the functionnality of the "compute_custom_only" flag? If yes, I'll set up a new (and cleaner) PR tested with my latest install of the package.

**Comment by doutriaux1** [comment]
Hum... That's a good question, I'm not sure, why use this at all if you only want YOUR computations? But I will let @durack1 and @gleckler1 decide on this.

**Comment by jservonnat** [comment]
Well, I think it depends on what you want to do with the MP. Today, the MP allows to compute the PCMDI metrics. It also cleverly allows the implementation of custom metrics. Following this, the MP can be used to compute all sorts of metrics. The point is that once you've computed the PCMDI metrics, you don't want to compute them again and again. I found very useful and practical to organize the computation of my different sets of metrics with different parameter files; one for each set of metrics. For example, I firstly compute the PCMDI on all the variables. Then, I run the MP to compute only the latitude of the jets.
With this flag, I think that the MP could be more interesting to compute custom metrics and have a broader use.
Up to you guys. It's just a couple of code lines and I can keep them for a personal version of the MP.

**Comment by durack1** [comment]
I personally like the idea that you can customize the package, and @jservonnat has implemented his custom functionality.. If @doutriaux1 sees this adaptation of the package as being something that is clean enough and general enough to build on, I suggest we keep it..

@gleckler1 opinions?

**Comment by durack1** [comment]
@jservonnat this has been deprecated by #265 so I'm closing

**Pull request closed without merging**